src/Protocol.cpp

```diff
@@ -0,0 +1,290 @@
+            0,
+            0xff);
+    }
+}
+
+std::optional<base::samples::Pointcloud> Protocol::handleSingleEcho(unsigned char* data)
+{
+    if (m_full_frame == true) {
+        m_point_cloud.points.clear();
+        m_point_cloud.colors.clear();
+        m_full_frame = false;
+    }
+    uint64_t timestamp_microseconds =
+        (static_cast<uint64_t>(data[1200] << 24) +
+            static_cast<uint64_t>(data[1201] << 16) +
+            static_cast<uint64_t>(data[1202] << 8) + data[1203]);
+    int miliseconds = timestamp_microseconds / 1000;
+    int microseconds = timestamp_microseconds % 1000;
+    base::Time time = base::Time::fromTimeValues(m_configuration.time_utc.year,
+        m_configuration.time_utc.month,
+        m_configuration.time_utc.day,
+        data[1197],
+        data[1198],
+        data[1199],
+        miliseconds,
+        microseconds);
+    for (int i = 0; i < 1197; i = i + 7) {
+        if (data[i] == 0xff && data[i + 1] == 0xaa && data[i + 2] == 0xbb &&
+            data[i + 3] == 0xcc && data[i + 4] == 0xdd) {
+            m_full_frame = true;
+            // TODO: Use the syncronized time here. We are using the time::now() at this
+            // moment cause we need a hardware integration between the lidar and a $GPRMC
+            // sender
+            m_point_cloud.time = base::Time::now();
+            return m_point_cloud;
+        }
+        if (data[i] < 128) {
+            double distance =
+                (data[i + 3] * 65536 + data[i + 4] * 256 + data[i + 5]) / 256.0 / 100;
+            if (distance != 0) {
+                uint8_t line_number = data[i];
+                double horizontal_angle_degree =
+                    (data[i + 1] * 256 + data[i + 2]) / 100.f;
+
+                uint8_t intensity = data[i + 6];
+
+                auto point = getPoint(line_number, horizontal_angle_degree, distance);
+                m_point_cloud.points.push_back(point);
+                m_point_cloud.colors.push_back(colorByReflectivity(intensity));
+            }
+        }
+    }
+    return std::nullopt;
+}
+
+double toRad(double degree)
+{
+    return (degree * M_PI / 180);
+}
+
+double Protocol::computeVerticalAngleSin(uint8_t line_number,
+    double horizontal_angle_degree)
+{
+    double r = m_cos_theta_2[line_number] * m_cos_theta_1[line_number] *
+                   cos(toRad(horizontal_angle_degree / 2.0)) -
+               m_sin_theta_2[line_number] * m_sin_theta_1[line_number];
+    double sin_vertical_angle =
+        m_sin_theta_1[line_number] + 2 * r * m_sin_theta_2[line_number];
+    return sin_vertical_angle;
+}
+
+void Protocol::computeDefaultSinesAndCosines()
+{
+    for (int i = 0; i < 128; i++) {
+        m_sin_theta_1[i] = sin(toRad(EMISSION_ANGLES_DEGREE[i / 4]));
+        m_sin_theta_2[i] = sin(toRad((i % 4) * (-0.17)));
+        m_cos_theta_1[i] = cos(toRad(EMISSION_ANGLES_DEGREE[i / 4]));
+        m_cos_theta_2[i] = cos(toRad((i % 4) * (-0.17)));
```

Comment:
I would prefer that you used base::Angle in here as well, allowing you to remove toRad

Reply:
I think it is less efficient to instantiate a base::Angle struct every time

Reply:
We don't need to use base::Angle every time we use a angle inside the code.